File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/python/06. Alien Dictonary.py
"""
Question:
Given a sorted dictionary of an alien language having N words and K starting alphabets,
find the order of characters in the alien language.

Approach:
- Compare adjacent words to determine character ordering (directed edges).
- Use Kahn's algorithm (topological sort) to determine the full order.

Time Complexity: O(N)
Space Complexity: O(K)
"""

from collections import deque


def set_adj(a, b, adj, indeg):
    n = min(len(a), len(b))
    for i in range(n):
        if a[i] != b[i]:
            u = ord(a[i]) - ord('a')
            v = ord(b[i]) - ord('a')
            adj[u].append(v)
            indeg[v] += 1
            break


def find_order(dictionary, N, K):
    adj = [[] for _ in range(K)]
    indeg = [0] * K

    for i in range(1, N):
        set_adj(dictionary[i - 1], dictionary[i], adj, indeg)

    queue = deque()
    for i in range(K):
        if indeg[i] == 0:
            queue.append(i)

    ans = ""
    while queue:
        node = queue.popleft()
        ans += chr(node + ord('a'))
        for v in adj[node]:
            indeg[v] -= 1
            if indeg[v] == 0:
                queue.append(v)

    return ans

File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/javascript/06. Alien Dictonary.js
/*
Question:
Given a sorted dictionary of an alien language having N words and K starting alphabets,
find the order of characters in the alien language.

Approach:
- Compare adjacent words to determine character ordering (directed edges).
- Use Kahn's algorithm (topological sort) to determine the full order.

Time Complexity: O(N)
Space Complexity: O(K)
*/

function setAdj(a, b, adj, indeg) {
    let n = Math.min(a.length, b.length);
    for (let i = 0; i < n; i++) {
        if (a[i] !== b[i]) {
            let u = a.charCodeAt(i) - 97;
            let v = b.charCodeAt(i) - 97;
            adj[u].push(v);
            indeg[v]++;
            break;
        }
    }
}

function findOrder(dict, N, K) {
    let adj = Array.from({ length: K }, () => []);
    let indeg = new Array(K).fill(0);

    // Set the directed edges and indegrees
    for (let i = 1; i < N; i++) {
        setAdj(dict[i - 1], dict[i], adj, indeg);
    }

    let queue = [];
    for (let i = 0; i < K; i++) {
        if (indeg[i] === 0) queue.push(i);
    }

    let ans = "";
    while (queue.length > 0) {
        let next = [];
        for (let node of queue) {
            ans += String.fromCharCode(node + 97);
            for (let v of adj[node]) {
                indeg[v]--;
                if (indeg[v] === 0) next.push(v);
            }
        }
        queue = next;
    }

    return ans;
}

